pyccel/parser/semantic.py

```diff
@@ -3766,17 +3894,55 @@ def _visit_Return(self, expr):
             expr  = Return(results)
         return expr
 
-    def _visit_FunctionDef(self, expr):
-        name            = self.scope.get_expected_name(expr.name)
-        decorators      = expr.decorators
-        funcs           = []
-        sub_funcs       = []
-        func_interfaces = []
-        is_pure         = expr.is_pure
-        is_elemental    = expr.is_elemental
-        is_private      = expr.is_private
-        is_inline       = expr.is_inline
-        docstring      = self._visit(expr.docstring) if expr.docstring else expr.docstring
+    def _visit_FunctionDef(self, expr, annotate=False, function_call=None):
+        """
+        Annotate the FunctionDef if necessary.
+
+        The FunctionDef is only annotated if the flag annotate is set to True.
+        In the case of an inlined function, we always annotate the function partialy,
+        depending on the function call if it is an interface, otherwise we annotate it
+        if the function_call argument are compatible with the FunctionDef arguments.
+        In the case of non inlined function, we only pass through this method
+        twice, the first time we do nothing and the second time we annotate all of functions.
+
+        Parameter
+        ---------
+        expr : FunctionDef|Interface
+           The that needs to be annotated.
+           If we provide an Interface, this means that the function has been annotated partialy,
+           and we need to continue annotating the needed ones.
+
+        annotate : bool, default: False
+           Annotate expr if the flag is set to True.
+
+        function_call: list, optional
+            The list of call argument, needed only in the case of an inlined function.
```

Comment:
Is this a list of `FunctionCall` nodes, or a list of `FunctionCallArgument` objects? In the latter case I recommend that we change the variable name to `function_call_args`, and that we change the word `argument` with the plural `arguments` in the docstring.

Reply:
[681b62d](https://github.com/pyccel/pyccel/pull/1728/commits/681b62dbc14f0f67b29d47d90e8869e1d9ad03f5)

Reply:
[422a2ec](https://github.com/pyccel/pyccel/pull/1728/commits/422a2ec655109f56bffe458f143347f4dcce9eba)